src/client/core/aad/auth-provider.ts

```diff
@@ -63,45 +79,80 @@ export default class AuthProvider {
                     { ...authRequest, prompt: "none" }
                 );
                 code = await authCodeCallback(url, true);
-            } catch (e) {
+            } catch (silentAuthExcpetion) {
                 url = await client.getAuthCodeUrl(authRequest);
                 code = await authCodeCallback(url);
             }
 
             const result: AuthorizationResult =
                 await client.acquireTokenByCode({ ...authRequest, code });
-            if (result) {
-                this._accounts[tenantId] = result.account;
-            }
             return result;
         }
     }
 
-    public logout(): void {
-        this._removeAccount();
+    public async logout(): Promise<void> {
+        this._logoutPromise = new Promise(resolve => {
+            this._removeAccounts().then(resolve);
+        });
     }
 
     protected _getClient(tenantId: string): ClientApplication {
         if (tenantId in this._clients) {
             return this._clients[tenantId];
         }
-        return this._clients[tenantId] = new PublicClientApplication({
+        const client = new PublicClientApplication({
             auth: {
                 clientId: this.config.clientId,
                 authority:
                     `${this.app.properties.azureEnvironment.aadUrl}${tenantId}/`
+            },
+            cache: {
+                cachePlugin: this._cachePlugin
             }
         });
+        if (!this._primaryClient) {
+            this._primaryClient = client;
+        }
+        this._clients[tenantId] = client;
+        return client;
     }
 
-    private async _removeAccount(): Promise<void> {
-        for (const tenantId in this._clients) {
-            if (this._accounts[tenantId]) {
-                const cache = this._clients[tenantId].getTokenCache();
-                cache.removeAccount(this._accounts[tenantId]);
+    private async _removeAccounts(): Promise<void> {
+        const cache = this._primaryClient?.getTokenCache();
```

Comment:
This can be undefined. Not sure why Typescript isn't complaining here (maybe we're using permission compiler flags)